students/arun_nalla/session06/test_Mailroom_Part4.py

```diff
@@ -0,0 +1,49 @@
+#!usr/bin/env python
+""" Testing Mailroom functions using pytest"""
+
+from Mailroom import get_report
+from Mailroom import donors_dict
+from Mailroom import donors_list
+from Mailroom import add_donors
+from Mailroom import email_message
+from Mailroom import path_to_store
+from Mailroom import add_donation
+import pytest
+import random
+import os
+
+demo_dict = {'Abc':[1, 2], 'Xyz':[10]}
+
+def test_donors_list():
+    actual = list(demo_dict)
+    assert actual == donors_list(demo_dict)
+
+def test_add_donors():
+    name = 'Arun'
+    assert add_donors(name, demo_dict) == demo_dict.setdefault(name,[])
+    assert 'Arun' in demo_dict
+    assert len(demo_dict['Arun']) == 0
+    assert 'ArunN' not in demo_dict
+
+
+def test_add_donations():
+    name = 'Arun'
+    donation = 1
+    expected_dict = {'Abc':[1, 2], 'Xyz':[10], name:[donation]}
+    assert add_donation(name, demo_dict, donation) == expected_dict
+    assert 1 in demo_dict['Arun']
+    assert len(demo_dict['Arun']) == 1
+    assert len(demo_dict['Abc']) == 2
+    assert 1 not in demo_dict['Xyz']
+    assert 10 not in demo_dict['Abc']
+
+def test_get_report():
+    expected_sorting = [('Xyz', 10, 1, 10.0), ('Abc', 3, 2, 1.5), ('Arun', 1, 1, 1.0)]
+    assert get_report(demo_dict) == expected_sorting
```

Comment:
Minor: The success of this test case depends on the execution of `test_add_donations()` (otherwise it'll fail), and I'm not sure how pytest guarantees that. In fact, I think it's safe not to assume such a specific ordering.

Probably the right approach is to initialize the `demo_dict` whenever a test case is executed and let the test case work on the re-initialized `demo_dict`. In fact, there are pytest constructs that allow such initialization (like xunit-style setup/teardown construct, or pytest fixtures), but we didn't cover them in class. At least we can have a separate function that initializes needed variables that can be called by every test case at the beginning. Just FYI for the future and no need to change your code.